src/topology/algebra/module.lean

```diff
@@ -371,31 +416,42 @@ rfl
   f₁.prod_map f₂ x = (f₁ x.1, f₂ x.2) :=
 rfl
 
-end general_ring
+/-- The continuous linear map given by `(x, y) ↦ f₁ x + f₂ y`. -/
+def coprod [topological_add_monoid M₃] (f₁ : M →L[R] M₃) (f₂ : M₂ →L[R] M₃) :
+  (M × M₂) →L[R] M₃ :=
+⟨linear_map.coprod f₁ f₂, (f₁.cont.comp continuous_fst).add (f₂.cont.comp continuous_snd)⟩
 
-section comm_ring
-
-variables
-{R : Type*} [comm_ring R] [topological_space R]
-{M : Type*} [topological_space M] [add_comm_group M]
-{M₂ : Type*} [topological_space M₂] [add_comm_group M₂]
-{M₃ : Type*} [topological_space M₃] [add_comm_group M₃]
-[module R M] [module R M₂] [module R M₃] [topological_module R M₃]
+@[norm_cast, simp] lemma coe_coprod [topological_add_monoid M₃]
+  (f₁ : M →L[R] M₃) (f₂ : M₂ →L[R] M₃) :
+  (f₁.coprod f₂ : (M × M₂) →ₗ[R] M₃) = linear_map.coprod f₁ f₂ :=
+rfl
 
-instance : has_scalar R (M →L[R] M₃) :=
-⟨λ c f, ⟨c • f, continuous_const.smul f.2⟩⟩
+@[simp] lemma coprod_apply [topological_add_monoid M₃] (f₁ : M →L[R] M₃) (f₂ : M₂ →L[R] M₃) (x) :
+  f₁.coprod f₂ x = f₁ x.1 + f₂ x.2 := rfl
 
-variables (c : R) (h : M₂ →L[R] M₃) (f g : M →L[R] M₂) (x y z : M)
+/-- Given a right inverse `f₂ : M₂ →L[R] M` to `f₁ : M →L[R] M₂`,
+`proj_ker_of_right_inverse f₁ f₂ h` is the projection `M →L[R] f₁.ker` along `f₂.range`. -/
+def proj_ker_of_right_inverse [topological_add_group M] (f₁ : M →L[R] M₂) (f₂ : M₂ →L[R] M)
+  (h : function.right_inverse f₂ f₁) :
+  M →L[R] f₁.ker :=
+(id R M - f₂.comp f₁).cod_restrict f₁.ker $ λ x, by simp [h (f₁ x)]
 
-@[simp] lemma smul_comp : (c • h).comp f = c • (h.comp f) := rfl
+@[simp] lemma coe_proj_ker_of_right_inverse_apply [topological_add_group M]
+  (f₁ : M →L[R] M₂) (f₂ h x) :
```

Comment:
Spelling out the types explicitly (at least of `f₂` and `h`) is of not use to Lean, but it might be of some use to your human readers :) Same thing in the next few lemmas. If you want, you can factor out `f₁` and `f₂` as section assumptions, but please keep the right inverse assumption in the theorems.

Reply:
Done.